external/esp_idf_port/esp32/esp32_queue_api.c

```diff
@@ -0,0 +1,434 @@
+			mq_close(queue_info->mqd_fd_recv);
+			queue_info->mqd_fd_recv = NULL;
+		}
+		queue_info->mq_item_size = 0;
+		queue_info->valid = false;
+	}
+	return;
+}
+
+/****************************************************************************
+ * Name: queue_send_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_timedsend to adds the specificied message (msg) to
+ * the message queue (mqdes).
+ *
+ * Inputs:
+ *	queue - The queue to send messages.
+ *	item - The message to be send.
+ *	block_time_tick - the time (in Ticks) to wait until a timeout is decleared
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+static bool queue_is_full(void *queue)
+{
+	queue_info_t *queue_info = (queue_info_t *) queue;
+	return false;
+}
+
+int32_t IRAM_ATTR queue_send_wrapper(void *queue, void *item, uint32_t block_time_tick)
+{
+	int32_t ret;
+	queue_info_t *queue_info = NULL;
+	struct timespec abstime;
+	uint32_t secs;
+	uint32_t msecs;
+	uint32_t nsecs;
+	uint32_t carry;
+
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	queue_info = (queue_info_t *) queue;
+	if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+		if (block_time_tick == 0xFFFFFFFF) {
+			ret = mq_send(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL);
+			if (ret == ERROR) {
+				return pdFAIL;
+			}
+		}
+		else {
+			(void)clock_gettime(CLOCK_REALTIME, &abstime);
+			msecs = TICK2MSEC(block_time_tick);
+			secs = msecs / MSEC_PER_SEC;
+			nsecs = (msecs - (secs * MSEC_PER_SEC)) * NSEC_PER_MSEC;
+			abstime.tv_sec += secs;
+			abstime.tv_nsec += nsecs;
+			if (abstime.tv_nsec >= NSEC_PER_SEC) {
+				carry = abstime.tv_nsec / NSEC_PER_SEC;
+				abstime.tv_sec += carry;
+				abstime.tv_nsec -= (carry * NSEC_PER_SEC);
+			}
+			ret = mq_timedsend(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL, &abstime);
+			if (ret == ERROR) {
+				return pdFAIL;
+			}
+		}
+	} else {
+		return pdFAIL;
+	}
+	return pdPASS;;
+}
+
+/****************************************************************************
+ * Name: queue_send_from_isr_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_send to adds the specificied message (msg) to
+ * the message queue (mqdes).
+ *
+ * Inputs:
+ *	queue - The queue to send messages.
+ *	item - The message to be send.
+ *	hptw - useless in TizenRT, but needed in the declaration of esp32 wifi os adapter.
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+int32_t IRAM_ATTR queue_send_from_isr_wrapper(void *queue, void *item, void *hptw)
+{
+	int32_t ret;
+	queue_info_t *queue_info = NULL;
+
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	queue_info = (queue_info_t *) queue;
+	if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+		ret = mq_send(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL);
```

Comment:
is it possible to use mq_send in isr ?  it will block if the queue is full, that's dangerous to block in isr

Reply:
`O_NONBLOCK` will be checked in `mq_waitsend()`/`mq_waitreceive()`

Reply:
mq_send  itself support call both in  task and ISR context
pls review code.